File: day5/day5.ts
const fs = require('fs');
import { Dictionary, every, filtermap, groupbytoobject, map, pipe, sum } from "powerseq";

interface MappedInput {
  rules: Dictionary<number[]>;
  numbers: number[][];
}

interface NumberWithCorrectness {
  n: number;
  correct: boolean;
}

type PageNumber = NumberWithCorrectness | number;

function parseInput(input: string): MappedInput {
  const [rules, numbers] = input.split('\n\n');
  return {
    rules: pipe(
      rules.split('\n'),
      map(r => r.split('|')),
      groupbytoobject(r => parseInt(r[0]), r => parseInt(r[1]))
    ),
    numbers: numbers.split('\n').map(r => r.split(',').map(Number))
  }
}

function getNumValue(num: PageNumber) {
  return typeof num === "number" ? num : num.n;
}

function isRuleFulfilled(row: PageNumber[], index: number, rule: number[]) {
  return pipe(
    row.slice(0, index),
    every(n => !rule.includes(getNumValue(n)))
  )
}

function isCorrectPage(page: PageNumber[], rules: Dictionary<number[]>) {
  if (!page) return;
  return page.every((num, i) => {
    const rule = rules[getNumValue(num)];
    return !rule || isRuleFulfilled(page, i, rule);
  });
}

//1.
function part1(input: MappedInput) {
  return pipe(
    input.numbers,
    filtermap(page => isCorrectPage(page, input.rules) ? page : null),
    sum(p => p[Math.round((p.length - 1) / 2)])
  )
}

//2.
function checkCorrectness(page: PageNumber[], rules: Dictionary<number[]>): NumberWithCorrectness[] {
  return page.map((num, i) => {
    const nVal = getNumValue(num);
    const rule = rules[nVal];
    return { n: nVal, correct: !rule || isRuleFulfilled(page, i, rule) };
  });
}

function swapWithPrevious(arr: PageNumber[], index: number) {
  if (index > 0) [arr[index - 1], arr[index]] = [arr[index], arr[index - 1]];
  return arr;
}

function makePageCorrect(page: NumberWithCorrectness[], rules: Dictionary<number[]>) {
  const firstIncorrect = page.findIndex(el => !el.correct);
  const correctedPage = checkCorrectness(swapWithPrevious(page, firstIncorrect), rules);
  return isCorrectPage(correctedPage, rules) ? correctedPage : makePageCorrect(correctedPage, rules);
}

function part2(input: MappedInput) {
  return pipe(
    input.numbers,
    filtermap(page => {
      const mapped = checkCorrectness(page, input.rules);
      return mapped.some(n => !n.correct) ? mapped : null
    }),
    map((page) => makePageCorrect(page, input.rules)),
    sum(p => p[Math.round((p.length - 1) / 2)].n)
  )
}

const input = parseInput(fs.readFileSync('./day5/input.txt', 'utf-8'));

console.log(part1(input));
console.log(part2(input));

File: day5/day5.go
package day5

import (
	"aoc2024/aochelpers"
	"strings"
	"math"
	"fmt"
	"sort"
)

type MappedInput struct {
	Rules   map[int][]int
	Numbers [][]int
}

func parseRules(rules string) map[int][]int{
	ruleMap := make(map[int][]int)
	for _, ruleLine := range strings.Split(rules, "\n") {
		ruleParts := strings.Split(ruleLine, "|")
		key := aochelpers.ParseInt(strings.TrimSpace(ruleParts[0]))
		value := aochelpers.ParseInt(strings.TrimSpace(ruleParts[1]))
		ruleMap[key] = append(ruleMap[key], value)
	}
	return ruleMap
}

func parseNumbers(numbers string) [][]int{
	var numberList [][]int
	for _, numberLine := range strings.Split(numbers, "\n") {
		numberParts := strings.Split(numberLine, ",")
		var numbersRow []int
		for _, numStr := range numberParts {
			num := aochelpers.ParseInt(strings.TrimSpace(numStr))
			numbersRow = append(numbersRow, num)
		}
		numberList = append(numberList, numbersRow)
	}
	return numberList
}

func parseInput(input string) MappedInput{
	parts := strings.Split(input, "\n\n")
	ruleMap := parseRules(parts[0])
	numberList := parseNumbers(parts[1])

	return MappedInput{
		Rules:   ruleMap,
		Numbers: numberList,
	}
}

func isFulfilledRule(row, rule []int, index int) bool{
	for i := 0; i < index; i++ {
		for _, rNum := range rule {
			if(row[i] == rNum){
				return false
			}
		}
	}
	return true
}

func isCorrectPage(page []int, rules map[int][]int, part2 bool) bool {
	for i, num := range page {
		rule, exists := rules[num]
		if exists {
			fulfilled := isFulfilledRule(page, rule, i)
			if !fulfilled {
				return false
			}
		}
	}
	return true
}

func sortPageCorrectly(page []int, rules map[int][]int) {
	sort.Slice(page, func(i, j int) bool {
		rule, exists := rules[i]
		if !exists {
			return true
		}
		for _, n := range rule {
			if n == j {
				return false
			}
		}
		return true
	})
}

func getMiddleIndexValue(page []int) int {
	middleIndex := int(math.Round(float64(len(page)-1) / 2))
	return page[middleIndex]
}

func calculateResult(numbers [][]int, rules map[int][]int, part2 bool) int {
	result := 0
	for _, page := range numbers {
		isCorrect := isCorrectPage(page, rules, part2)
		if isCorrect {
			if(!part2){
				result += getMiddleIndexValue(page)
			}
		} else if(part2){
			//tutaj pomysł od MN
			sortPageCorrectly(page, rules)
			result += getMiddleIndexValue(page)
		}
	}
	return result
}

func Part1() {
	input := parseInput(aochelpers.ReadFileAsString("./day5/input.txt"))
	result := calculateResult(input.Numbers, input.Rules, false)

	fmt.Println("Day 5, part 1:", result)
}

func Part2() {
	input := parseInput(aochelpers.ReadFileAsString("./day5/input.txt"))
	result := calculateResult(input.Numbers, input.Rules, true)

	fmt.Println("Day 5, part 2:", result)
}


